### ISSUE TITLE
Color manipulation in templates

### ISSUE THREAD

**Issue by adrusi** [deviantfero/wpgtk#154]
It would be very useful for templates to support color manipulation functions that could be used to, for instance, alter the brightness or saturation of a color from the theme, blend colors from the theme, etc.

### Specific Use-Case

I have a vscode theme that gets generated from wpgtk colors. I'd like for the sidebar to be a similar color to the editor background. Currently I can do this by setting the sidebar background color to be semi-transparent:

![2019-06-06 15:46:21](https://user-images.githubusercontent.com/183309/59061785-45c99980-8872-11e9-8bf2-4f7d30bc034a.png)

This works because it turns out that underneath the sidebar, there's a black background, and setting the sidebar background opacity to less than one results in blending it with black (i.e. reducing its brightness). But that doesn't work with light themes — blending the color with black achieves the opposite of the desired effect, making the dark text on the sidebar less legible. Ideally I'd want to blend the editor background color with the active color, but that can't be achieved with opacity.

Thus I'm left with a poor delineation between the sidebar and editor when using a light theme:

![2019-06-06 15:38:53](https://user-images.githubusercontent.com/183309/59061583-d0f65f80-8871-11e9-80ad-b4b1815f8e65.png)

Supporting functions for blending (and otherwise manipulating) colors in templates would solve this problem.

### Suggested Implementation

Within template braces, allow function call syntax, and support color manipulation functions like in [Sass](https://sass-lang.com/documentation/functions/color). Example:

```json
{{
    "name": "vs-wal",
    "type": "dark",
    "colors": {{
        "sideBar.background": "{mix(background, active, 75%)}"
    }}
}}
```

Parentheses within template braces would be an error in the current template syntax, so this is fully backwards compatible. All old templates would continue to work.

### Disadvantages

Color themes look good when there aren't too many different colors on the screen. Wpgtk currently helps guide users away from this failure mode by limiting the total number of colors to 16. This feature would make it easier to make ugly themes.

---

I'm happy to implement this feature if it's desired.

**Comment by deviantfero** [comment]
hey @adrusi, I would love to have this feature, but I'm afraid it might break compatibility between `pywal` and `wpgtk` templates I would like them to keep them easily transferable, we can make clever use of the format in the templates to achieve many things though, like you said opacity is achievable and if the file in question allows it, you can even mix colors too, right now it mostly depends on the software that interprets the template rather than the template itself.

one solution to your vscode colorscheme I might suggest is using `{color8}` for the sidebar and `{color0}` for the actual background, if you've auto-adjusted your colorschemes with `wpgtk` it should provide a nice result almost every time.

I use this technique in Thunar for the GTK+ theme and it seems to work good enough, take a look

[thunar-white](https://imgur.com/tfav4Ke)
[thunar-black](https://imgur.com/Mg7HP19)

I am interested though in knowing what you had in mind for achieving this.

**Comment by adrusi** [comment]
`{color8}` works well for the sidebar. I had to adjust what color8 is defined as in my colorschemes, but it doesn't look like that broke anything. I don't auto-adjust my colorschemes, I leave them as color***n+8*** = color***n***, not for any particular reason. I don't think I use wpgtk as recommended (although I find it useful anyway) so it's good to know that my templates can be reasonably compatible with other setups if I assume color8 is different but similar to color0.

I no longer have a specific use-case for color manipulation features in the template language. Given that this is a power-user's tool, it might still be good to include them, since someone is bound to find a use for them. If _bidirectional_ compatibility with pywal is necessary, then the only way to make it work would be to change pywal's template engine as well.

My vscode theme can be found in [my dotfiles repo](https://github.com/adrusi/dotfiles/tree/a7ac47ebd03ce3d4a02494dc71a5488f6e36042c/vscode). It's based on [Bluedrack28's vscode-wal](https://github.com/Bluedrack28/vscode-wal), and takes all of the syntax highlighting definitions from there, but with very different UI theming. I'll send him a PR now that I've got it working a bit nicer.

As far as including it in the repo, that would be cool! Vscode doesn't really like being themed by an external program though, so the setup involves side-loading two extensions (vscode decides whether to use light or dark icons based on whether the themes name ends in `-light` as far as I can tell), and then configuring wpgtk to modify the theme files in the extensions. I'm not sure how to best package that up portably.
